src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexParser.cs

```diff
@@ -0,0 +1,1927 @@
+        private RegexToken _currentToken;
+        private int _recursionDepth;
+
+        private RegexParser(
+            ImmutableArray<VirtualChar> text, RegexOptions options,
+            ImmutableDictionary<string, TextSpan> captureNamesToSpan,
+            ImmutableDictionary<int, TextSpan> captureNumbersToSpan) : this()
+        {
+            _lexer = new RegexLexer(text);
+            _options = options;
+
+            _captureNamesToSpan = captureNamesToSpan;
+            _captureNumbersToSpan = captureNumbersToSpan;
+
+            // Get the first token.  It is allowed to have trivia on it.
+            ConsumeCurrentToken(allowTrivia: true);
+        }
+
+        /// <summary>
+        /// Returns the latest token the lexer has produced, and then asks the lexer to 
+        /// produce the next token after that.
+        /// </summary>
+        /// <param name="allowTrivia">Whether or not trivia is allowed on the next token
+        /// produced.  In the .net parser trivia is only allowed on a few constructs,
+        /// and our parser mimics that behavior.  Note that even if trivia is allowed,
+        /// the type of trivia that can be scanned depends on the current RegexOptions.
+        /// For example, if <see cref="RegexOptions.IgnorePatternWhitespace"/> is currently
+        /// enabled, then '#...' comments are allowed.  Otherwise, only '(?#...)' comments
+        /// are allowed.</param>
+        private RegexToken ConsumeCurrentToken(bool allowTrivia)
+        {
+            var previous = _currentToken;
+            _currentToken = _lexer.ScanNextToken(allowTrivia, _options);
+            return previous;
+        }
+
+        /// <summary>
+        /// Given an input text, and set of options, parses out a fully representative syntax tree 
+        /// and list of diagnostics.  Parsing should always succeed, except in the case of the stack 
+        /// overflowing.
+        /// </summary>
+        public static RegexTree TryParse(ImmutableArray<VirtualChar> text, RegexOptions options)
+        {
+            try
+            {
+                // Parse the tree once, to figure out the capture groups.  These are needed
+                // to then parse the tree again, as the captures will affect how we interpret
+                // certain things (i.e. escape references) and what errors will be reported.
+                //
+                // This is necessary as .net regexes allow references to *future* captures.
+                // As such, we don't know when we're seeing a reference if it's to something
+                // that exists or not.
+                var tree1 = new RegexParser(text, options,
+                    ImmutableDictionary<string, TextSpan>.Empty,
+                    ImmutableDictionary<int, TextSpan>.Empty).ParseTree();
+
+                var analyzer = new CaptureInfoAnalyzer(text);
+                var (captureNames, captureNumbers) = analyzer.Analyze(tree1.Root, options);
```

Comment:
Since we don't need to hold onto the analyzer, I'd replace the constructor and `Analyze` call with a call to a new private method. Then the constructor and `Analyze` can be made private and there is no risk of calling one without calling the other (ie. leak).

Reply:
I'm not sure i understand what this means.  Can you clarify?